### DISCUSSION TITLE
mixed debug between JavaScript and Rust.

### DISCUSSION THREAD

**Discussion by bvanjoi** [web-infra-dev/rspack#4389] [Show and tell]
This discussion aims to illustrate the method for mixed debugging between JavaScript and Rust.

## Prerequisites

To illustrate this process, I'll use an example. Let's start by introduce the environment and example I have used.

- System: macos
- IDE: vscode
- Debugging target: `rspack build ${projectRoot}/examples/basic`

Firstly, you need to build rspack in debug mode. To do this, execute the following commands in the project's root directory:

```bash
npm run build:binding:debug
npm run build:js
```

## Configure `launch.json` in vscode

It's necessary to configure two debug configurations within ﻿`.vscode/launch.json.`

- attach for node:

```jsonc
{
  "name": "attach:node”,
  "request": "attach",  // refer: https://code.visualstudio.com/docs/editor/debugging#_launch-versus-attach-configurations
  "type": "node",
  // `9229` is the default port of message
  "port": 9229
}
```

- and launch for lldb

```jsonc
{
  "name": "launch:rust-from-node",
  "request": "launch”,
  "type": "lldb",    // it means we use `lldb` to launch the binary file of `node`
  "program": "node”,
  "args": [
    "--inspect",
    "--enable-source-maps",
    "${workspaceFolder}/packages/rspack-cli/bin/rspack",
    "build",
    "-c",
    "${workspaceFolder}/examples/basic/rspack.config.js",
  ],
   // `cwd` is just for repack find the correclty entry.
  "cwd": "${workspaceFolder}/examples/basic/"
}
```

Next, we can utilize [compounds](https://code.visualstudio.com/docs/editor/debugging#_compound-launch-configurations) to amalgamate the two commands:

```json
{
  "name": "mix-debug",
  "configurations": [
    "attach:node",
    "launch:rust-from-node"
  ]
}
```

Finally, your `﻿launch.json` should appear as follows:

```json
{
  "configurations": [
    {
      "name": "attach:node",
      "request": "attach",
      "type": "node",
      "port": 9229
    },
    {
      "name": "launch:rust-from-node",
      "request": "launch",
      "type": "lldb",
      "program": "node",
      "args": [
        "--inspect",
        "--enable-source-maps",
        "${workspaceFolder}/packages/rspack-cli/bin/rspack",
        "build",
        "-c",
        "${workspaceFolder}/examples/basic/rspack.config.js",
      ],
      "cwd": "${workspaceFolder}/examples/basic/"
    }
  ],
  "compounds": [
    {
      "name": "mix-debug",
      "configurations": [
        "attach:node",
        "launch:rust-from-node"
      ]
    }
  ]
}
```

## Debugging Attempt

Next, we can introduce some breakpoints and commence debugging. 

The result appears as follows:

https://github.com/web-infra-dev/rspack/assets/30187863/106983f7-a59e-4d9e-9001-552f4441d88b

**Comment by hardfist** [comment]
really nice explanation

**Comment by shuaiqidekaige** [comment]
不知为何，在我的电脑上无法正常断点，总是自动结束

**Reply by hardfist** [reply]
你是编译的debug版本吗

**Reply by shuaiqidekaige** [reply]
按照贡献文档中的指南，我操作了下来，在debug环节，按照上述操作，始终无法进入断点，vscode版本是1.83.1

**Reply by shuaiqidekaige** [reply]
https://github.com/web-infra-dev/rspack/assets/31154493/2c0ae2d5-1408-4d33-8a89-302dbf77b43c

**Reply by bvanjoi** [reply]
- I recommend expressing your issue in English. This approach allows the information to be accessible to a broader audience.
- Instead of only providing a screen recording of the failed debug attempt, please offer more context. For instance, including the build step in your recording can be very helpful.
- It could be beneficial to use ﻿`node --inspect-brk` to confirm a successful attachment of the node debugger.

**Reply by shuaiqidekaige** [reply]
Okay, here are my steps。
1. pnpm install and cargo build
2. pnpm run build:cli:debug
3. use the provided lauch.json file
4. begin debug

**Reply by shuaiqidekaige** [reply]
just these operations show that the build is successful on the terminal, but the breakpoint cannot be entered.

**Reply by bvanjoi** [reply]
Your steps appear to be correct.

 Could you please try replacing `--inspect` with `--inspace-brk` in the launch file and retry again?

```json
"name": "launch:rust-from-node",
"request": "launch",
"type": "lldb",
"program": "node",
"args": [
      "--inspect", // change it to `--inspace-brk`
```

**Reply by shuaiqidekaige** [reply]
I tried it, and it seems to only take effect in node, and cannot be located in the ts file. and the breakpoints in rust still don’t take effect.

**Reply by shuaiqidekaige** [reply]
https://github.com/web-infra-dev/rspack/assets/31154493/5b2ebc7b-f8fd-4ce2-948b-7182bde8ee97

**Reply by bvanjoi** [reply]
> cannot be located in the ts file

This issue arises from the absence of a sourcemap for TypeScript. However, it's not something you need to worry about.

> the breakpoints in rust still don’t take effect

Could you please set a breakpoint at the entry point of Rust, for example, at `﻿node_binding::Rspack::new`

**Reply by shuaiqidekaige** [reply]
https://github.com/web-infra-dev/rspack/assets/31154493/d5421910-d222-439b-863c-539f36f68ff1

**Reply by bvanjoi** [reply]
I suspect that the output of rust is in production mode. Could you please clean the output of the binding first, then rerun ﻿`npm run build:cli:debug`? The easiest way to clean is by deleting this project and re-cloning it.

**Reply by shuaiqidekaige** [reply]
I tried re-pulling the project and made sure to use `npm run build:cli:debug`，but it still doesn't work。I think this is a problem with vscode debug。

**Reply by h-a-n-a** [reply]
Would you try again with the latest configuration of `launch.json` in main?
